python/pylibvw.cc

```diff
@@ -688,31 +691,32 @@ void ex_push_feature_dict(example_ptr ec, vw_ptr vw, unsigned char ns, PyObject*
   PyObject *key, *value;
   Py_ssize_t key_size = 0, pos = 0;
   float feat_value;
-  uint64_t feat_index;
+  feature_index feat_index;
 
   while (PyDict_Next(o, &pos, &key, &value))
   {
-    if (PyFloat_Check(value)) { feat_value = (float)PyFloat_AsDouble(value); }
-    else if (PyLong_Check(value))
-    {
-      feat_value = (float)PyLong_AsDouble(value);
-    }
+    if (PyLong_Check(value)) { feat_value = (float)PyLong_AsDouble(value); }
     else
     {
-      std::cerr << "warning: malformed feature in list" << std::endl;
-      continue;
+      feat_value = (float)PyFloat_AsDouble(value);
+      if (feat_value == -1 && PyErr_Occurred())
```

Comment:
Why this check?

Reply:
If `value` can't be turned into a float for some reason then this will get flagged (e.g., if `value` is a string). This is equivalent to the check on line 635 in ex_push_feature_list. In my benchmarks I found that doing the check this way was a tiny bit faster than trying to do error checking ourselves before calling `PyFloat_AsDouble`.

Reply:
Oh also, in case what you're asking is why the strange check for a value of -1 and then `PyErr_Occured`, [here's a link](https://docs.python.org/3/c-api/float.html#c.PyFloat_AsDouble) to the documentation. You can see that error handling is performed by checking for a value of -1 and then calling `PyErr_Occurred`. I'm assuming they do it this way for performance reasons, though I don't know.

Reply:
Cool thanks for the link

Reply:
Yeah, in my testing I've found that using the CPython functions directly is quite a bit faster than using the BoostPython methods, which is why I switched from the more readable boost methods to the CPython methods which are a little more cryptic.